File: language/src/main/java/brj/nodes/LocalVarNode.java
package brj.nodes;

import brj.BridjeLanguage;
import com.oracle.truffle.api.dsl.NodeField;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.frame.FrameSlot;
import com.oracle.truffle.api.frame.FrameSlotTypeException;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.source.SourceSection;
import org.jetbrains.annotations.NotNull;

import static com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere;

@NodeField(name = "frameSlot", type = FrameSlot.class)
@NodeField(name = "sourceSection", type = SourceSection.class)
public abstract class LocalVarNode extends ExprNode {
    public LocalVarNode(BridjeLanguage lang) {
        super(lang);
    }

    public abstract FrameSlot getFrameSlot();

    @NotNull
    @Specialization
    public Object doExecute(@NotNull VirtualFrame frame) {
        try {
            return frame.getObject(getFrameSlot());
        } catch (FrameSlotTypeException e) {
            throw shouldNotReachHere(e);
        }
    }
}

File: language/src/main/kotlin/brj/nodes/LocalVarNode.kt
package brj.nodes

import brj.BridjeLanguage
import com.oracle.truffle.api.CompilerDirectives
import com.oracle.truffle.api.dsl.Specialization
import com.oracle.truffle.api.frame.FrameSlotTypeException
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.instrumentation.StandardTags
import com.oracle.truffle.api.instrumentation.Tag
import com.oracle.truffle.api.source.SourceSection

abstract class LocalVarNode(lang: BridjeLanguage, loc: SourceSection?, private val frameSlotIdx: Int) : ExprNode(lang, loc) {
    @Specialization
    fun doExecute(frame: VirtualFrame): Any {
        return try {
            frame.getAuxiliarySlot(frameSlotIdx)
        } catch (e: FrameSlotTypeException) {
            throw CompilerDirectives.shouldNotReachHere(e)
        }
    }

    override fun hasTag(tag: Class<out Tag>): Boolean {
        return tag == StandardTags.ReadVariableTag::class.java || super.hasTag(tag)
    }
}